source/extensions/clusters/aggregate/cluster.h

```diff
@@ -0,0 +1,156 @@
+#pragma once
+
+#include "envoy/config/cluster/aggregate/cluster.pb.h"
+#include "envoy/config/cluster/aggregate/cluster.pb.validate.h"
+
+#include "common/upstream/cluster_factory_impl.h"
+#include "common/upstream/upstream_impl.h"
+
+#include "extensions/clusters/aggregate/lb_context.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Clusters {
+namespace Aggregate {
+
+using PriorityContext = std::pair<Upstream::PrioritySetImpl,
+                                  std::vector<std::pair<uint32_t, Upstream::ThreadLocalCluster*>>>;
+
+class Cluster : public Upstream::BaseDynamicClusterImpl, Upstream::ClusterUpdateCallbacks {
+public:
+  Cluster(const envoy::api::v2::Cluster& cluster,
+          const envoy::config::cluster::aggregate::ClusterConfig& config,
+          Upstream::ClusterManager& cluster_manager, Runtime::Loader& runtime,
+          Runtime::RandomGenerator& random,
+          Server::Configuration::TransportSocketFactoryContext& factory_context,
+          Stats::ScopePtr&& stats_scope, ThreadLocal::SlotAllocator& tls, bool added_via_api);
+
+  // Upstream::Cluster
+  Upstream::Cluster::InitializePhase initializePhase() const override {
+    return Upstream::Cluster::InitializePhase::Secondary;
+  }
+
+  // Upstream::ClusterUpdateCallbacks
+  void onClusterAddOrUpdate(Upstream::ThreadLocalCluster& cluster) override;
+  void onClusterRemoval(const std::string& cluster_name) override;
+
+  void refresh() {
+    refresh([](const std::string&) { return false; });
+  }
+
+  std::vector<std::string> clusters_;
+  Upstream::ClusterUpdateCallbacksHandlePtr handle_;
+  Upstream::ClusterManager& cluster_manager_;
+  Runtime::Loader& runtime_;
+  Runtime::RandomGenerator& random_;
+  ThreadLocal::SlotPtr tls_;
+
+private:
+  // Upstream::ClusterImplBase
+  void startPreInit() override;
+
+  void refresh(const std::function<bool(const std::string&)>& skip_predicate);
+  PriorityContext
+  linearizePrioritySet(const std::function<bool(const std::string&)>& skip_predicate);
+};
+
+// Load balancer used by each worker thread. It will be refreshed when clusters, hosts or priorities
+// are updated.
+class AggregateClusterLoadBalancer : public Upstream::LoadBalancer {
+public:
+  AggregateClusterLoadBalancer(Upstream::ClusterStats& stats, Runtime::Loader& runtime,
+                               Runtime::RandomGenerator& random,
+                               const envoy::api::v2::Cluster::CommonLbConfig& common_config)
+      : stats_(stats), runtime_(runtime), random_(random), common_config_(common_config) {}
+
+  // Upstream::LoadBalancer
+  Upstream::HostConstSharedPtr chooseHost(Upstream::LoadBalancerContext* context) override;
+
+private:
+  class LoadBalancerImpl : public Upstream::LoadBalancerBase {
```

Comment:
hmm i'm still not convinced we need this. can we just have the top level LB extend LBBase?

Reply:
If we have the top level LB extend LBBase, it needs to be initialized when the factory calls create, it's possible that when we initialize the top level LB, the priority set is empty. Using empty priority set to initialize LBBase will trigger segfault. (It tries to access empty array with index 0). In this case, we should set the inner LB as `nullptr`.

Reply:
OK that makes sense to me. Mind leaving a comment explaining the reasoning for future readers?

Reply:
Comment added.